src/integer_mod_q/mat_polynomial_ring_zq/concat.rs

```diff
@@ -0,0 +1,247 @@
+    fn concat_horizontal(self, other: Self) -> Result<Self::Output, crate::error::MathError> {
+        if self.get_num_rows() != other.get_num_rows() {
+            return Err(MathError::MismatchingMatrixDimension(format!(
+                "Tried to concatenate horizontally a '{}x{}' matrix and a '{}x{}' matrix.",
+                self.get_num_rows(),
+                self.get_num_columns(),
+                other.get_num_rows(),
+                other.get_num_columns()
+            )));
+        }
+        let mut matrix = MatPolyOverZ::new(
+            self.get_num_rows(),
+            self.get_num_columns() + other.get_num_columns(),
+        )
+        .unwrap();
+        unsafe {
+            fmpz_poly_mat_concat_horizontal(
+                &mut matrix.matrix,
+                &self.matrix.matrix,
+                &other.matrix.matrix,
+            );
+        }
+        Ok(MatPolynomialRingZq {
+            matrix,
+            modulus: self.get_mod(),
+        })
+    }
+}
+
+#[cfg(test)]
+mod test_concatenate {
+    use crate::{
+        integer::MatPolyOverZ,
+        integer_mod_q::{MatPolynomialRingZq, ModulusPolynomialRingZq, PolyOverZq},
+        traits::{Concatenate, GetNumColumns, GetNumRows},
+    };
+    use std::str::FromStr;
+
+    const BITPRIME64: u64 = u64::MAX - 58;
+
+    /// Ensure that the dimensions are taken over correctly and an error occurs
+    /// if the dimensions mismatch
+    #[test]
+    fn dimensions_vertical() {
+        let poly_mod =
+            PolyOverZq::from_str(&format!("3  1 {} 1 mod {}", i64::MAX, BITPRIME64)).unwrap();
+        let modulus = ModulusPolynomialRingZq::try_from(&poly_mod).unwrap();
+        let mat_1 = MatPolynomialRingZq::new(13, 5, &modulus).unwrap();
+        let mat_2 = MatPolynomialRingZq::new(17, 5, &modulus).unwrap();
+        let mat_3 = MatPolynomialRingZq::new(17, 6, &modulus).unwrap();
+
+        let mat_vert = mat_1.concat_vertical(&mat_2).unwrap();
+
+        assert_eq!(5, mat_vert.get_num_columns());
+        assert_eq!(30, mat_vert.get_num_rows());
+        assert!(mat_1.concat_vertical(&mat_3).is_err());
+    }
+
+    /// Ensure that the dimensions are taken over correctly and an error occurs
+    /// if the dimensions mismatch
+    #[test]
+    fn dimensions_horizontal() {
+        let poly_mod =
+            PolyOverZq::from_str(&format!("3  1 {} 1 mod {}", i64::MAX, BITPRIME64)).unwrap();
+        let modulus = ModulusPolynomialRingZq::try_from(&poly_mod).unwrap();
+        let mat_1 = MatPolynomialRingZq::new(13, 5, &modulus).unwrap();
+        let mat_2 = MatPolynomialRingZq::new(17, 5, &modulus).unwrap();
+        let mat_3 = MatPolynomialRingZq::new(17, 6, &modulus).unwrap();
+
+        let mat_vert = mat_2.concat_horizontal(&mat_3).unwrap();
+
+        assert_eq!(11, mat_vert.get_num_columns());
+        assert_eq!(17, mat_vert.get_num_rows());
+        assert!(mat_1.concat_horizontal(&mat_2).is_err());
```

Comment:
```suggestion
        let mat_vert = mat_2.concat_horizontal(&mat_3).unwrap();
        assert!(mat_1.concat_horizontal(&mat_2).is_err());

        assert_eq!(11, mat_vert.get_num_columns());
        assert_eq!(17, mat_vert.get_num_rows());
```
A bit hard to spot the error assert. Since it calls `concat_horizontal`, you can also move it up.